Sources/VRMSceneKit/CustomType/VRMSpringBone.swift

```diff
@@ -0,0 +1,201 @@
+                localPosition.x * scale.x,
+                localPosition.y * scale.y,
+                localPosition.z * scale.z
+            ))
+            self.verlet.append(logic)
+        }
+
+        for child in parent {
+            self.setupRecursive(center, child)
+        }
+    }
+    
+    private func setLocalRotationsIdentity() {
+        for verlet in self.verlet {
+            verlet.head.utx.localRotation = SCNQuaternion.identity
+        }
+    }
+    
+    func update(deltaTime: TimeInterval) {
+        if self.verlet.isEmpty {
+            if self.rootBones.isEmpty {
+                return
+            }
+            setup()
+        }
+
+        self.colliderList = []
+        for group in self.colliderGroups {
+            for collider in group.colliders {
+                self.colliderList.append(SphereCollider(
+                    position: group.node.utx.transformPoint(collider.offset),
+                    radius: collider.radius
+                ))
+            }
+        }
+
+        let stiffness = self.stiffnessForce * SCNFloat(deltaTime)
+        let external = self.gravityDir * (self.gravityPower * SCNFloat(deltaTime))
+
+        for verlet in self.verlet {
+            verlet.radius = self.hitRadius
+            verlet.update(
+                center: self.center,
+                stiffnessForce: stiffness,
+                dragForce: self.dragForce,
+                external: external,
+                colliders: self.colliderList)
+        }
+    }
+}
+
+extension VRMSpringBone {
+    class VRMSpringBoneLogic {
+        let node: SCNNode
+        public var head: SCNNode { self.node }
+        private let length: SCNFloat
+        private var currentTail: SCNVector3
+        private var prevTail: SCNVector3
+        private let localRotation: SCNQuaternion
+        private let boneAxis: SCNVector3
+        private var parentRotation: SCNQuaternion {
+            self.node.parent?.utx.rotation ?? SCNQuaternion.identity
+        }
+        var radius: SCNFloat = 0.5
+        
+        init(center: SCNNode, node: SCNNode, localChildPosition: SCNVector3) {
+            self.node = node
+            let worldChildPosition = node.utx.transformPoint(localChildPosition)
+            self.currentTail = center.utx.inverseTransformPoint(worldChildPosition)
+            self.prevTail = self.currentTail
+            self.localRotation = node.utx.localRotation
+            self.boneAxis = localChildPosition.normalized
+            self.length = localChildPosition.length
+        }
+        
+        func update(center: SCNNode, stiffnessForce: SCNFloat, dragForce: SCNFloat, external: SCNVector3, colliders: [SphereCollider]) {
+            let currentTail: SCNVector3 = center.utx.transformPoint(self.currentTail)
+            let prevTail: SCNVector3 = center.utx.transformPoint(self.prevTail)
+
+            // verlet積分で次の位置を計算
```

Comment:
I didn't know Unity's and UniVRM's spring bone are implemented by the Verlet physics.
Thanks for the knowledge sharing. 😄

Reply:
This comment bring from original code.
https://github.com/vrm-c/UniVRM/blob/master/Assets/VRM/UniVRM/Scripts/SpringBone/VRMSpringBone.cs#L122

I could implement logic without expert knowledge thanks to original code.  🤫

Reply:
And this is a nice implementation in Swift 👍👍👍